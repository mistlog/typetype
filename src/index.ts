export * from "./parser";
export * from "./generator";
export * from "./core";