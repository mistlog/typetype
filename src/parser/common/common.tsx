import { ReactPeg } from "react-peg";

// white space
export function _() {
    return (
        <repeat type="*">
            <set> \t\n\r</set>
        </repeat>
    );
}

export const Text = (content: string) => {
    return (
        <list>
            <_ /><text>{content}</text><_ />
        </list>
    )
}