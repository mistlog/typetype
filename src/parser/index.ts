export * from "./common";
export * from "./basic";
export * from "./function";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";