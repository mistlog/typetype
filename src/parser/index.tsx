export * from "./common";
export * from "./basic";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";