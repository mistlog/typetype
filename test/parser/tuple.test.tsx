import { ReactPeg } from "react-peg";
import { TupleType, OperatorType } from "../../src";
import { saveAST } from "../common";

describe("parser: tuple", () => {
    test("TupleType", () => {
        const parser = ReactPeg.render(<TupleType />);
        const ast = parser.parse(`[protocol, rest]`);
        saveAST(ast, "TupleType.json");
        expect(ast).toMatchSnapshot();
    })

    test("TupleType: rest", () => {
        const parser = ReactPeg.render(<TupleType />);
        const ast = parser.parse(`[protocol, ...rest]`);
        saveAST(ast, "TupleType-Rest.json");
        expect(ast).toMatchSnapshot();
    })

    test("TupleType: readonly", () => {
        const parser = ReactPeg.render(<OperatorType />);
        const ast = parser.parse(`readonly [string, number, object]`);
        saveAST(ast, "TupleType-Readonly.json");
        expect(ast).toMatchSnapshot();
    })

    test("TupleType: with infer", () => {
        const parser = ReactPeg.render(<TupleType />);
        const ast = parser.parse(`[infer protocol, infer rest]`);
        saveAST(ast, "TupleType-Infer.json");
        expect(ast).toMatchSnapshot();
    })
})
