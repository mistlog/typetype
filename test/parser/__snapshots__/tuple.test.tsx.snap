// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser: tuple TupleType 1`] = `
Object {
  "items": Array [
    Object {
      "kind": "TypeReference",
      "typeName": Object {
        "kind": "Identifier",
        "name": "protocol",
      },
    },
    Object {
      "kind": "TypeReference",
      "typeName": Object {
        "kind": "Identifier",
        "name": "rest",
      },
    },
  ],
  "kind": "TupleType",
}
`;

exports[`parser: tuple TupleType: readonly 1`] = `
Object {
  "kind": "ReadonlyTuple",
  "operand": Object {
    "items": Array [
      Object {
        "kind": "StringType",
        "value": "string",
      },
      Object {
        "kind": "NumberType",
        "value": "number",
      },
      Object {
        "kind": "ObjectType",
        "value": "object",
      },
      Object {
        "kind": "BigIntType",
        "value": "bigint",
      },
    ],
    "kind": "TupleType",
  },
}
`;

exports[`parser: tuple TupleType: rest 1`] = `
Object {
  "items": Array [
    Object {
      "kind": "TypeReference",
      "typeName": Object {
        "kind": "Identifier",
        "name": "protocol",
      },
    },
    Object {
      "kind": "RestType",
      "param": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "rest",
        },
      },
    },
  ],
  "kind": "TupleType",
}
`;

exports[`parser: tuple TupleType: with infer 1`] = `
Object {
  "items": Array [
    Object {
      "kind": "InferType",
      "typeName": Object {
        "kind": "Identifier",
        "name": "protocol",
      },
    },
    Object {
      "kind": "InferType",
      "typeName": Object {
        "kind": "Identifier",
        "name": "rest",
      },
    },
  ],
  "kind": "TupleType",
}
`;
